### PR TITLE
[FLINK-24139][table-planner] Push down more predicates through Join in stream mode

### PR DESCRIPTION
**Pull request by xuyangzhong**
## What is the purpose of the change

Enable JoinDependentConditionDerivationRule in stream mode.

## Brief change log

  - *add a new rule set in FlinkStreamRuleSets*
  - *add a new program in FlinkStreamProgram*
  - *add some test case in stream/table/JoinTest, stream/sqlJoinTest, stream/table/JoinItCase, stream/sql/JoinItCase*

## Verifying this change

Some test cases added in stream/table/JoinTest, stream/sqlJoinTest, stream/table/JoinItCase and stream/sql/JoinItCase can verify this change.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented?

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkStreamRuleSets.scala [hunk 1]
```diff
@@ -159,6 +159,12 @@ object FlinkStreamRuleSets {
     CoreRules.FILTER_MERGE
   )
 
+  /**
+   * RuleSet to extracts sub-condition which can be pushed into join inputs
```
Comment (zjuwangg) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkStreamRuleSets.scala:163]:
```suggestion
   * RuleSet to extract sub-condition which can be pushed into join inputs
```

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkStreamProgram.scala [hunk 2]
```diff
@@ -121,11 +121,20 @@ object FlinkStreamProgram {
...
+                .setHepMatchOrder(HepMatchOrder.BOTTOM_UP)
+                .add(FlinkStreamRuleSets.JOIN_PREDICATE_REWRITE_RULES)
+                .build(), "join predicate rewrite")
+            .addProgram(
+              FlinkHepRuleSetProgramBuilder.newBuilder
+                .setHepRulesExecutionType(HEP_RULES_EXECUTION_TYPE.RULE_COLLECTION)
+                .setHepMatchOrder(HepMatchOrder.BOTTOM_UP)
+                .add(FlinkStreamRuleSets.FILTER_PREPARE_RULES)
+                .build(), "filter rules")
+            .setIterations(5).build(), "predicate rewrite")
```
Comment (zjuwangg) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkStreamProgram.scala:137]: Why we need set iterations and why chose 5 as the iteration number?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/table/JoinTest.scala [hunk 3]
```diff
@@ -31,6 +31,48 @@ import java.sql.Timestamp
   */
 class JoinTest extends TableTestBase {
 
+  @Test
+  def testDependentConditionderivationInnerJoin: Unit = {
```
Comment (zjuwangg) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/table/JoinTest.scala:35]: Ditto and all new added following method name should change.

**Comment by xuyangzhong** [comment]
Hi, @zjuwangg  
The main reason is that in FlinkBatchProgram, this iterator time is set with 5. In order to be consistent, i rewrite the number 5 here.
I tried to find why we need to set iterations here. Actually, I think this aims to pushdown the filter condition into the join condition and can continue optimizing the join condition several times. 
For example, in table api, the api named 'join' and 'where' will create a join relNode with a true condition and a fliter above the join relNode. The real condition we need will be set in the filter relNode.  The optimized program will do JOIN_PREDICATE_REWRITE_RULES first, which will extract the join condition, and then do FILTER_PREPARE_RULES, which will push the filter condition into the join condition. However, if we don't set the iterator, JOIN_PREDICATE_REWRITE_RULES finds that the join condition is always True, and will do nothing, which is not in our expectation. I reversed the order of executing these two rules and get the result i expected. I think the iterator time which means the repeat execution times will help avoid this kind of problem by repeating executing these two rules many times, which even can improve the problem of recursive optimization.

**Review by zjuwangg — APPROVED** [review]

**Comment by zjuwangg** [comment]
cc @godfreyhe to have another look

**Review by godfreyhe — APPROVED** [review]
Thanks for the contribution @xuyangzhong , and for the review @zjuwangg , LGTM, I will merge it

**Pull request closed without merging**